pkg/clusteragent/admission/mutate/auto_instrumentation.go

```diff
@@ -189,24 +201,88 @@ func injectApmTelemetryConfig(pod *corev1.Pod) {
+// - otherwise returns all tracing libraries supported by APM Instrumentation
+func getLibrariesToInjectForApmInstrumentation(pod *corev1.Pod, registry string) ([]libInfo, bool) {
+	autoDetected := false
+
+	// Pinned tracing libraries in APM Instrumentation configuration
+	libsToInject := getPinnedLibraries(registry)
+	if len(libsToInject) > 0 {
+		return libsToInject, autoDetected
+	}
+
+	// Tracing libraries for language detection
+	libsToInject = getLibrariesLanguageDetection(pod, registry)
+	if len(libsToInject) > 0 {
+		autoDetected = true
+		return libsToInject, autoDetected
+	}
+
+	// Latest tracing libraries for all supported languages (java, js, dotnet, python, ruby)
+	libsToInject = getAllLatestLibraries(registry)
+
+	return libsToInject, autoDetected
+}
+
+// getPinnedLibraries returns tracing libraries to inject as configured by apm_config.instrumentation.lib_versions
+func getPinnedLibraries(registry string) []libInfo {
+	// check if pinned libraries have been cached
+	pinnedLibs.libsMutex.Lock()
+	libsToInject := pinnedLibs.libraries
+	pinnedLibs.libsMutex.Unlock()
+
+	if libsToInject != nil {
+		return libsToInject
+	}
+
+	// get pinned libraries and cache them
 	var libVersion string
 	singleStepLibraryVersions := config.Datadog.GetStringMapString("apm_config.instrumentation.lib_versions")
+	if len(singleStepLibraryVersions) == 0 {
+		return []libInfo{}
+	}
+	libs := []libInfo{}
 
-	for _, lang := range supportedLanguages {
-		libVersion = "latest"
-
-		if version, ok := singleStepLibraryVersions[string(lang)]; ok {
-			log.Warnf("Library version %s is specified for language %s", version, string(lang))
-			libVersion = version
+	// If APM Instrumentation is enabled and configuration apm_config.instrumentation.lib_versions specified, inject only the libraries from the configuration
+	for lang, version := range singleStepLibraryVersions {
+		if !slices.Contains(supportedLanguages, language(lang)) {
+			log.Warnf("APM Instrumentation detected configuration for unsupported language: %s. Tracing library for %s will not be injected", lang, lang)
+			continue
 		}
+		log.Infof("Library version %s is specified for language %s", version, lang)
+		libVersion = version
+		libs = append(libs, libInfo{lang: language(lang), image: libImageName(registry, language(lang), libVersion)})
+	}
+	pinnedLibs.libsMutex.Lock()
+	defer pinnedLibs.libsMutex.Unlock()
+	pinnedLibs.libraries = libs
 
-		libsToInject[string(lang)] = libInfo{
-			lang:  lang,
-			image: libImageName(registry, lang, libVersion),
-		}
+	return libs
+}
+
+// getLibrariesLanguageDetection runs process language auto-detection and returns languages to inject for APM Instrumentation.
+// The langages information is available in workloadmeta-store and attached on the pod's owner.
+func getLibrariesLanguageDetection(pod *corev1.Pod, registry string) []libInfo {
+	libsToInject := getAutoDetectedLibraries(pod, registry)
```

Comment:
Shouldn't be this under the `if`?
Otherwise we'll do unnecessary work in that function if `inject_auto_detected_libraries` is false.

Reply:
You are right. I thought for a minute that getAutoDetectedLibraries triggers annotation of the Deployments.